### PR TITLE
Stratified Couette flow verification

### LINKED ISSUE #310: Stratified plane Couette flow test
We agreed that stratified plane Couette flow could be a test case against which we can verify the implementation of AMD in PR #309, although the published results [Vreugdenhil & Taylor (2018)] use a slightly modified version of AMD.

They report a number of LES runs, and since we don't have a vertically stretched grid, it might be easiest to try and reproduce the Pr = 0.7 case which they do with (Nx, Ny, Nz) = (64, 49, 64) grid points [for Ri=0 or 0.01, for Ri = 0.04 it's (64, 65, 64)]. I might suggest focusing on the Ri = 0 case as they also run a resolved DNS with (Nx, Ny, Nz) = (256, 129, 256) against which we might be able to compare (might not be possible, depending on vertical grid stretching).

Unfortunately, I don't think we can reproduce their results without a vertically stretched grid... If I understand the paper correctly, the grid is stretched according to
```
y_j = h*tanh(Sf*(2*(j-1)/(Ny-1))) / tanh(Sf)
```
so for h=1 you get grid spacings of Δy ~ 0.125 away from the wall and Δy ~ 0.0000035 adjacent to the wall. So a faithful reproduction would need 285,000+ vertical levels lol.

I don't fully understand how they got their values for the vertical grid cell size adjacent to the wall `∆y_w^+`, but it seems like if h = 100,000 then our values agree (they get a spacing of Δy ~ 0.35 adjacent to the wall). Either way, the ratio between the thickest and thinnest spacings is ~35,000.

Not sure if there's still a way we could compare results in this case without a vertically stretched grid...

Absolute worst case scenario, I've been working on an implementation of a vertically stretched grid (see PRs #283 and #306) but it's a work in progress and might not be the best use of our efforts right now.

Reference: Catherine A. Vreugdenhil and John R. Taylor, [Large-eddy simulations of stratified plane Couette flow using the anisotropic minimum-dissipation model](https://aip.scitation.org/doi/abs/10.1063/1.5037039), Physics of Fluids 30, 085104 (2018).

cc @glwagner @rafferrari

### LINKED ISSUE #415: We should use CI to test that examples and verification tests run without errors

### PR DESCRIPTION
**Pull request by ali-ramadhan**
This PR introduces a script for simulating Stratified Couette flow with the aim of reproducing the LES results of Vreugdenhil & Taylor (2018). See https://github.com/climate-machine/Oceananigans.jl/issues/310 for more details.

When the verification is completely successfully, I will ensure the script is maintained in the `test/verification/` subdirectory so that the verification can be reproduced at any time. I will also convert it to a regression test. See #347.

Resolves #310 
Resolves #415
Contributes to #346 and #347

### CODE REVIEW

File: test/verification/stratified_couette_flow.jl [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+wbcs = HorizontallyPeriodicBCs(    top = BoundaryCondition(Value, 0),
+                                bottom = BoundaryCondition(Value, 0))
+
+# Non-dimensional model setup
+model = Model(
+         arch = HAVE_CUDA ? GPU() : CPU(),
+            N = (128, 128, 1024),
+            L = (4π*h, 2π*h, 2h),
+      # closure = ConstantIsotropicDiffusivity(ν=ν, κ=κ),
+      closure = VerstappenAnisotropicMinimumDissipation(ν=ν, κ=κ),
```
Comment (glwagner) [inline test/verification/stratified_couette_flow.jl]: needs to be just `AnisotropicMinimumDissipation` on current master

File: test/verification/stratified_couette_flow.jl [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+
+append!(model.diagnostics, [Up, Vp, Wp, Tp, νp, κp])
+
+profiles = Dict(
+     :u => model -> Array(Up.profile),
+     :v => model -> Array(Vp.profile),
+     :w => model -> Array(Wp.profile),
+     :T => model -> Array(Tp.profile),
+    :nu => model -> Array(νp.profile),
+:kappaT => model -> Array(κp.profile))
```
Comment (glwagner) [inline test/verification/stratified_couette_flow.jl]: whoa

File: test/verification/stratified_couette_flow.jl [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+using Oceananigans
+using Oceananigans.TurbulenceClosures
+
+""" Friction velocity squared. See equation (16) of Vreugdenhil & Taylor (2018). """
+function uτ²(model)
+    Δz = model.grid.Δz
+    ν = model.closure.ν
+
+    Up = HorizontalAverage(model, model.velocities.u; frequency=Inf)
+    U = Up(model)[2:end-1]
```
Comment (glwagner) [inline test/verification/stratified_couette_flow.jl]: Explicitly use halo size `model.grid.Hz` for clarity / ensured correctness here?

File: test/verification/stratified_couette_flow.jl [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+    uτ²⁺, uτ²⁻ = uτ²(model)
+    qw⁺, qw⁻ = qw(model)
+    uτ⁺, uτ⁻ = √uτ²⁺, √uτ²⁻
+
+    qw⁺ / uτ⁺, qw⁻ / uτ⁻
+end
+
+""" Obukov length scale (assuming a linear equation of state). See equation (17) of Vreugdenhil & Taylor (2018). """
+function L_Obukov(model)
+    kₘ = 0.4  # Von Kármán constant
```
Comment (glwagner) [inline test/verification/stratified_couette_flow.jl]: blessed Von Karman constant, where would we be without you.

**Comment by glwagner** [comment]
Is "verification" just a run script, or do we include analysis as well? Maybe need to have a folder for each verification case?

**Comment by ali-ramadhan** [comment]
Thanks for the comments!

> Is "verification" just a run script, or do we include analysis as well? Maybe need to have a folder for each verification case?

Yeah a folder for each might make sense. I'm envisioning that each verification test case is described in the documentation with enough figures/plots to convince the reader that Oceananigans is doing the right thing. There should be a simulation script (like the one in this PR) that allows the reader to reproduce the simulation, and probably one or more plotting scripts allowing them to recreate the plots.

File: test/verification/stratified_couette_flow.jl [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+
+""" Friction velocity squared. See equation (16) of Vreugdenhil & Taylor (2018). """
+function uτ²(model)
+    Δz = model.grid.Δz
+    ν = model.closure.ν
+
+    Up = HorizontalAverage(model, model.velocities.u; frequency=Inf)
+    U = Up(model)[2:end-1]
+
+    uτ²⁺ = ν * abs(U[1] - Uw) / Δz
```
Comment (glwagner) [inline test/verification/stratified_couette_flow.jl]: Isn't this wrong? The distance between the `u` points and the wall is `Δz/2`, right?

File: test/verification/stratified_couette_flow.jl [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+    uτ²⁻ = ν * abs(-Uw - U[end]) / Δz
```
Comment (glwagner) [inline test/verification/stratified_couette_flow.jl]: At first I thought this was a bug --- the velocity prescribed at the bottom wall is `-Uw` though, right? This may deserve a comment. Also due to our k-indexing convention, this looks like you are computing the negative of the gradient and then taking the absolute value. This seems mathematically correct, but is slightly confusing. Perhaps compute the positive gradient and then take the absolute value?

File: test/verification/stratified_couette_flow.jl [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+
+""" Heat flux at the wall. See equation (16) of Vreugdenhil & Taylor (2018). """
+function qw(model)
+    Δz = model.grid.Δz
+    κ = model.closure.κ
+
+    Tp = HorizontalAverage(model, model.tracers.T; frequency=Inf)
+    Θ = Tp(model)[2:end-1]
+
+    qw⁺ = κ * abs(Θ[1] - Θw) / Δz
```
Comment (glwagner) [inline test/verification/stratified_couette_flow.jl]: Ditto for comments above about the velocity gradients at the wall.

File: test/verification/stratified_couette_flow.jl [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+Tbcs = HorizontallyPeriodicBCs(    top = BoundaryCondition(Value,  Θw),
+                                bottom = BoundaryCondition(Value, -Θw))
+
+ubcs = HorizontallyPeriodicBCs(    top = BoundaryCondition(Value,  Uw),
+                                bottom = BoundaryCondition(Value, -Uw))
+
+vbcs = HorizontallyPeriodicBCs(    top = BoundaryCondition(Value, 0),
+                                bottom = BoundaryCondition(Value, 0))
+
+wbcs = HorizontallyPeriodicBCs(    top = BoundaryCondition(Value, 0),
```
Comment (glwagner) [inline test/verification/stratified_couette_flow.jl]: This is wrong. This boundary condition should be `NoPenetration`.
